src/datta_lab_to_nwb/markowitz_gillis_nature_2023/preconversion/extract_metadata.py

```diff
@@ -292,9 +323,11 @@ def get_session_name(session_df):
     session_name : str
         Session name.
     """
-    session_names = set(session_df.session_name[session_df.session_name.notnull()]) | set(
-        session_df.SessionName[session_df.SessionName.notnull()]
-    )
+    session_names = set(session_df.SessionName[session_df.SessionName.notnull()])
+    try:
+        session_names = session_names.union(set(session_df.session_name[session_df.session_name.notnull()]))
+    except AttributeError:  # No session name column
+        pass
     assert len(session_names) <= 1, "Multiple session names found"
     try:
         session_name = session_names.pop()
```

Comment:
Same here - recommend finding a way to avoid the `try/except`

Reply:
Don't get me wrong, sometimes try/except is the ONLY way to assert certain things, but this does not seem like one of those

The majority of the time a try/except is simply too broad in its catching that it can be harder to determine the resulting error from a situation than it is to predict if a bad input is the cause

It also helps make code more readable when the try clause has a bunch of stuff going on

Specifically in this situation, an `AttributeError` could occur in at least 3 places, though technically many more. The `session_names` variable may not have the bound method `union`. `session_df` may not have the attribute `session_name`. The attribute `session_df.session_name` may not have the bound method `notnull`. I'm assuming the result of `[session_df.session_name.notnull()]` would yield a `KeyError` if not what `session_df.session_name` can support indexing of, but that's also an assumption being made that may not always be 100% correct. And much worse than that, even if those 3 conditions are met, if ANY of the hundreds of lines of code that the bound methods call throughout the stack trace encounters an `AttributeError` for some other reason (which even if you don't think it can happen now, what if the upstream libraries release a change that does make it a possibility years down the road?) it will still count as leading to the except case here and will not even warn you about what happened when you run your program.

If any of these situations occurred for whatever seemingly silly reason, you may not be made aware of it because it entered the black hole of the blanket try/except - with normal `if/else`, all of those cases would have raised the specific error from the specific operation and you would have been given a traceback that could be used to fix the problem

At a minimum, some ways to improve would be (a) check that the actual message from the `AttributeError` matches your exact expectation of output and raise it again otherwise, (b) downgrade the error to a warning, or warn something more informative to your use case in the pipeline so we or the person running the code has some awareness of what has transpired.

But overall that seems like much more work than simply asking

```
if the_specific_condition_you_want_to_avoid_is_met:
    do_the_thing_you_want_to_do()
```

I've also never heard this phrase "easier to ask for forgiveness than permission" used in our circles, and would not recommend that attitude in general within our ecosystem or with our clientele - ALWAYS ask permission first and avoid all roads that lead to a situation that might require asking forgiveness, as that is a much worse situation to find yourself in.

If interested, you could ask other members of the team for their stories about how try/excepts have caused unforeseen problems in code if we have time in a future meeting

Reply:
Yes, I mostly agree with @CodyCBakerPhD and this has been our style and policy. This is mostly for the reason Cody stated, that exceptions tend to be too broad. As the article stated, it's also a problem when there are side effects to the code. We generally reserve try/except for when they are really necessary, which is helpful because that then holds meaning- if I see a try/except my first thought is that this is probably a situation where another approach would be impossible prohibitively awkward.

Thanks for linking the article though, @pauladkisson . It's helpful to see where your side is coming from. It taught me some ideas about the difference between these two different styles, and where EAFP might be better than LBYL in some cases. I think we could consider the EAFP style in some limited instances.

Reply:
Thank you both for taking the time to explain the reasoning behind this preference, I will keep it in mind going forward!